File: use-cases/easy-ship/code/java/src/main/java/lambda/utils/HttpFileTransferUtil.java
package lambda.utils;

import java.io.*;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.zip.GZIPInputStream;
import com.squareup.okhttp.*;

public class HttpFileTransferUtil {

    private static final OkHttpClient httpClient = new OkHttpClient();

    /**
     * Downloads a document from the specified URL and returns it as an InputStream.
     *
     * @param url The URL to download the document from.
     * @param compressionAlgorithm The compression algorithm to use (e.g., "GZIP").
     * @return An InputStream containing the downloaded document.
     * @throws IOException If there is an error during download.
     * @throws IllegalArgumentException If the character set or compression algorithm is invalid.
     */
    public static InputStream download(String url, String compressionAlgorithm) throws IOException, IllegalArgumentException {
        Request request = new Request.Builder()
                .url(url)
                .get()
                .build();

        Response response = httpClient.newCall(request).execute();
        if (!response.isSuccessful()) {
            String errorMessage = String.format("Call to download document failed with response code: %d and message: %s",
                    response.code(), response.message());
            throw new IOException(errorMessage);
        }

        ResponseBody responseBody = response.body();
        if (responseBody == null) {
            throw new IOException("Received an empty response body");
        }

        // Get the charset from Content-Type
        MediaType mediaType = MediaType.parse(response.header("Content-Type"));
        Charset charset = mediaType != null ? mediaType.charset(StandardCharsets.UTF_8) : StandardCharsets.UTF_8;

        InputStream inputStream = responseBody.byteStream();
        if ("GZIP".equalsIgnoreCase(compressionAlgorithm)) {
            inputStream = new GZIPInputStream(inputStream);
        }

        return inputStream;  // Returning the InputStream for the downloaded document
    }

    /**
     * Uploads the provided byte array to the specified URL.
     *
     * @param source The byte array of the document to upload.
     * @param url The URL to upload the document to.
     * @throws IOException If there is an error during upload.
     */
    public static void upload(byte[] source, String url) throws IOException {
        String contentType = String.format("text/xml; charset=%s", StandardCharsets.UTF_8);
        Request request = new Request.Builder()
                .url(url)
                .put(RequestBody.create(MediaType.parse(contentType), source))
                .build();

        Response response = httpClient.newCall(request).execute();
        if (!response.isSuccessful()) {
            String errorMessage = String.format("Call to upload document failed with response code: %d and message: %s",
                    response.code(), response.message());
            throw new IOException(errorMessage);
        }
    }
}

File: use-cases/easy-ship/code/php/lambda/utils/HttpFileTransferUtil.php
<?php

namespace Lambda\Utils;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\Psr7\Stream;
use RuntimeException;

class HttpFileTransferUtil
{
    private static ?Client $httpClient = null;

    /**
     * Initializes the HTTP client.
     */
    private static function getHttpClient(): Client
    {
        if (self::$httpClient === null) {
            self::$httpClient = new Client([
                'timeout' => 30.0, // Set a timeout for HTTP requests
            ]);
        }
        return self::$httpClient;
    }

    /**
     * Downloads a document from the specified URL and returns it as a stream.
     *
     * @param string $url The URL to download the document from.
     * @param string|null $compressionAlgorithm The compression algorithm to use (e.g., "GZIP").
     * @return \Psr\Http\Message\StreamInterface The downloaded document as a stream.
     * @throws RuntimeException|GuzzleException If the download fails or the response is invalid.
     */
    public static function download(
        string $url,
        ?string $compressionAlgorithm = null
    ): \Psr\Http\Message\StreamInterface {
        try {
            $response = self::getHttpClient()->get($url);

            if ($response->getStatusCode() !== 200) {
                throw new RuntimeException(sprintf(
                    'Call to download document failed with response code: %d and message: %s',
                    $response->getStatusCode(),
                    $response->getReasonPhrase()
                ));
            }

            $stream = $response->getBody();

            // Handle GZIP compression if specified
            if (strtolower($compressionAlgorithm) === 'gzip') {
                $gzippedData = $stream->getContents();
                $uncompressedData = gzdecode($gzippedData);
                if ($uncompressedData === false) {
                    throw new RuntimeException('Failed to decode GZIP compressed data.');
                }
                return \GuzzleHttp\Psr7\Utils::streamFor($uncompressedData);
            }

            return $stream;
        } catch (RequestException $e) {
            throw new RuntimeException('Error during document download: ' . $e->getMessage(), 0, $e);
        }
    }

    /**
     * Uploads the provided byte array to the specified URL.
     *
     * @param string $data The content of the document to upload.
     * @param string $url The URL to upload the document to.
     * @throws RuntimeException|GuzzleException If the upload fails.
     */
    public static function upload(string $data, string $url): void
    {
        try {
            $response = self::getHttpClient()->put($url, [
                'body' => $data,
                'headers' => [
                    'Content-Type' => 'text/xml; charset=UTF-8',
                ],
            ]);

            if ($response->getStatusCode() !== 200) {
                throw new RuntimeException(sprintf(
                    'Call to upload document failed with response code: %d and message: %s',
                    $response->getStatusCode(),
                    $response->getReasonPhrase()
                ));
            }
        } catch (RequestException $e) {
            throw new RuntimeException('Error during document upload: ' . $e->getMessage(), 0, $e);
        }
    }
}
